apps/labs/posts/numpy-thread-safe-test-suite.md

```diff
@@ -0,0 +1,333 @@
+
+### Random Number Generation
+
+The next biggest category of test failures involved the usage of `np.random` in tests.
+
+#### Problem
+
+Generally, anything that invokes global state and shared data will cause thread safety issues. And well, `np.random` very much is global state! When threads use `np.random`, they are using the same global RNG instance, which can cause test failures when tests rely on seeded results.
+
+Let's say you have a test that relies on seeded results, such as below.
+
+```python
+def test_rng():
+  np.random.seed(123)
+  x = np.random.rand()
+  y = np.random.rand()
+  z = np.random.rand()
+  ...
+```
+
+When ran normally under a single thread, you may get results like this for the three variables:
+
+| x     | y     | z     |
+| ----- | ----- | ----- |
+| 0.696 | 0.286 | 0.226 |
+
+However, when ran with multiple threads, the threads fight for the usage of the global RNG. So, you may instead get results like so:
+
+|          | x     | y     | z     |
+| -------- | ----- | ----- | ----- |
+| Thread 1 | 0.696 | 0.226 | 0.719 |
+| Thread 2 | 0.696 | 0.286 | 0.551 |
+
+You can see how the RNG seed is sort of shared between threads. Obviously, not a good thing when we're trying to keep things thread safe. Tests that care about seed results will fail since they're not getting the results they expect.
+
+#### Solution
+
+Thankfully, NumPy actually has a solution for us! Instead of using the global `np.random` instance, we can create local instances with `np.random.RandomState`.
+
+```python
+def test_rng():
+  rng = np.random.RandomState(123)
+  x = rng.rand()
+  y = rng.rand()
+  z = rng.rand()
+  ...
+```
+
+Another somewhat simple solution to a problem that affected large parts of the test suite. Nice!
+
+**Note:** NumPy also has the newer Generator class that handles local creation of RNG instances. However, RandomState utilizes the same RNG as `np.random`, making it so I didn't need to go in and modify the expected results.
+
+### Temporary Files
+
+And finally, the last major change I'll go in-depth with is the usage of temporary files.
+
+#### Problem
+
+As stated early, usage of the file system is thread unsafe, since the file system is very much global state. When it comes to making temporary files for test suites, it's commonplace to utilize fixtures is `pytest` that create temporary file paths, like `tmp_path`. If you're not careful with file names and locations, tests that use `tmp_path` can be thread-unsafe.
+
+#### Solution
+
+Overall, to make temporary file usage thread safe, we want to make sure file paths are unique between threads. One way we thought of early on was appending uuids at the end of file names. This did definitely work, but after further consideration, we decided to look elsewhere for a more foolproof solution.
```

Comment:
UUID is an acronym and should be all-caps. Maybe link to the UUID wikipedia article or just explain that UUIDv4 is a way to generate ASCII strings that are overwhelmingly statisticaly likely to be unique.